service/controller/resource/cproutetables/resource.go

```diff
@@ -66,43 +76,44 @@ func (r *Resource) addRouteTablesToContext(ctx context.Context) error {
 		return microerror.Mask(err)
 	}
 
-	// We check if we have all route tables cached for the configured route table
-	// names. If we find all information, we return them.
-	{
-		r.logger.LogCtx(ctx, "level", "debug", "message", "finding cached route tables")
+	if len(r.routeTables) == r.expectedTableCount && r.expectedTableCount > 0 {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "found cached route tables")
+		cc.Status.ControlPlane.RouteTables = r.routeTables
 
-		if len(r.routeTables) == len(r.names) {
-			r.logger.LogCtx(ctx, "level", "debug", "message", "found cached route tables")
-			cc.Status.ControlPlane.RouteTables = r.routeTables
-
-			return nil
-		} else {
-			r.logger.LogCtx(ctx, "level", "debug", "message", "did not find cached route tables")
-		}
+		return nil
 	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "did not find cached route tables")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "caching route tables")
+	if len(r.names) == 0 {
+		// We do not have the cached route tables, so we look them up using tags.
+		tables, err := r.lookupByTag(ctx, cc.Client.ControlPlane.AWS.EC2, r.installation)
+		if err != nil {
+			return microerror.Mask(err)
+		}
 
-	// We do not have the cached route tables, so we look them up.
-	{
-		r.logger.LogCtx(ctx, "level", "debug", "message", "caching route tables")
-
+		r.routeTables = tables
+	} else {
+		// We do not have the cached route tables, so we look them up using names
+		// supplied via RouteTables flag.
 		for _, name := range r.names {
-			rt, err := r.lookup(ctx, cc.Client.ControlPlane.AWS.EC2, name)
+			rt, err := r.lookupByName(ctx, cc.Client.ControlPlane.AWS.EC2, name)
 			if err != nil {
 				return microerror.Mask(err)
 			}
 
 			r.routeTables = append(r.routeTables, rt)
 		}
-
-		r.logger.LogCtx(ctx, "level", "debug", "message", "cached route tables")
 	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "cached route tables")
 
 	cc.Status.ControlPlane.RouteTables = r.routeTables
+	r.expectedTableCount = len(r.routeTables)
```

Comment:
Does it make sense to have that `r.expectedTableCount`? I feel like it's only testing if `len` returns the same value for the same slice.

Reply:
RouteTables don't change over time. If we get rid of `r.expectedTableCount`, we are hitting AWS API every time to get the same list of tables. This way we ask once and perform lookup again if something changes.

Reply:
I'm asking if we can replace:

```
if len(r.routeTables) == r.expectedTableCount && r.expectedTableCount > 0 {
```

with:

```
if len(r.routeTables) > 0 {
```

and get rid of `r.expectedTableCount`.

Reply:
Smart, I'll do that.